File: leet_code/trees/lowest_common_ancestor.js
// if lCA is bigger than both p and q so we move left
// if lCA is smaller than both p and q so we move right
// if it doesn't meet both checks return that node

var lowestCommonAncestor = (root, p, q) => {
    
    if(p.val < root.val && q.val < root.val){
        return lowestCommonAncestor(root.left,p,q);
    }

    if(p.val > root.val && q.val > root.val){
        return lowestCommonAncestor(root.right,p,q);
    }

    return root;
};

File: leet_code/python/binary_trees/lowest_common_ancestor.py

# O(n) time, O(h) space
def lca(root,node_one,node_two):
    if not root: return root
    if root.val == p.val or root.val == q.val: return root
    
    left = lca(root.left, p, q)
    right = lca(root.right, p, q)
    
    if left and right:
        return root
    elif not left and not right:
        return None
    else:
        return left or right